src/compiler/checker.ts

```diff
@@ -12563,15 +12564,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 case SyntaxKind.MappedType:
                 case SyntaxKind.ConditionalType: {
                     const outerTypeParameters = getOuterTypeParameters(node, includeThisTypes);
-                    if (node.kind === SyntaxKind.MappedType) {
+                    if ((kind === SyntaxKind.FunctionExpression || kind === SyntaxKind.ArrowFunction || isObjectLiteralMethod(node)) && isContextSensitive(node as Expression | MethodDeclaration)) {
+                        const signature = firstOrUndefined(getSignaturesOfType(getTypeOfSymbol(getSymbolOfDeclaration(node as FunctionLikeDeclaration)), SignatureKind.Call));
+                        if (signature && signature.typeParameters) {
+                            return [...(outerTypeParameters || emptyArray), ...signature.typeParameters];

```

Comment:
Should this be concatenate like the other cases?

Reply:
We are required to return a mutable array, and `concatenate` only returns a mutable array when both arguments are mutable arrays. In this case, `signature.typeParameters` is read-only.

Can't get rid of the empty array, spreading `undefined` causes an exception.

Reply:
I meant more that `concatenate` can accept `undefined` so you don't have to `||` things, but, yeah, it doesn't return a copy.

Though it's sort of confusing given all of the surrounding code uses `append` / `concatenate`